nilearn/image/image.py

```diff
@@ -502,6 +502,61 @@ def swap_img_hemispheres(img):
 
     # create swapped nifti object
     out_img = nibabel.Nifti1Image(img.get_data()[::-1], img.get_affine(),
-        header=img.get_header())
+                                  header=img.get_header())
 
     return out_img
+
+
+def _index_img(imgs, index):
+    """Helper function for index_img and iter_img.
+    """
+    return nibabel.Nifti1Image(imgs.get_data()[:, :, :, index],
+                               imgs.get_affine(),
+                               header=imgs.get_header())
+
+
+def index_img(imgs, index):
+    """Indexes into a 4D Niimg-like object in the fourth dimension.
+
+    Can be used to extract a 3D image out of `imgs` or create a 4D
+    image whose data is a subset of the `imgs` data.
+
+    Parameters
+    ----------
+    imgs: 4D Niimg-like object
+          See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+
+    index: Any type compatible with numpy array indexing
+           Used for indexing the 4D data array in the fourth dimension
+
+    Returns
+    -------
+    output: nibabel.Nifti1IMage
+            3D or 4D image depending on the index used
+
+    """
+    imgs = check_niimgs(imgs)
+    return _index_img(imgs, index)
+
+
+def iter_img(imgs):
+    """Iterates over a 4D Niimg-like object in the fourth dimension.
+
+    Parameters
+    ----------
+    imgs: 4D Niimg-like object
+          See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+
+    Returns
+    -------
+    output: iterator of 3D nibabel.Nifti1Image
+    """
+    class Checked4DImageIterator(object):
+        def __init__(self, imgs):
+            self.imgs = check_niimgs(imgs)
+
+        def __iter__(self):
+            for i in range(_get_shape(self.imgs)[3]):
+                yield _index_img(self.imgs, i)
```

Comment:
> FWIW, I seem to be able to pickle iter_img fine, even with the nested class

From my understanding of Python scopes, you shouldn't be able to pickle the result of iter_img (the iterator itself) but I can be wrong.

> The main reason I put it inside the class it that otherwise iter_img and Checked4DImageIterator are an alias of each other, which doesn't seem ideal.

I have no strong feeling. I just prefer classes outside of functions for the sake of readability. I honestly didn't think of people trying to pickle iterators.